Change device scan default accumulator type

umfranzw opened the pull request:
The default accumulator types are changing for ROCm 7.0 as described below. Note that this is a breaking change.
  * rocprim::inclusive_scan
    * previous default: class AccType = typename std::iterator_traits<InputIterator>::value_type>
    * new default: class AccType = rocprim::invoke_result_binary_op_t<typename std::iterator_traits<InputIterator>::value_type, BinaryFunction>`
  * `rocprim::deterministic_inclusive_scan
    * previous default: class AccType = typename std::iterator_traits<InputIterator>::value_type>`
    * new default: class AccType = rocprim::invoke_result_binary_op_t<typename std::iterator_traits<InputIterator>::value_type, BinaryFunction>`
  * `rocprim::exclusive_scan
    * previous default: class AccType = detail::input_type_t<InitValueType>>
    * new default: class AccType = rocprim::invoke_result_binary_op_t<rocprim::detail::input_type_t<InitValueType>, BinaryFunction>
  * `rocprim::deterministic_exclusive_scan
    * previous default: class AccType = detail::input_type_t<InitValueType>>
    * new default: class AccType = rocprim::invoke_result_binary_op_t<rocprim::detail::input_type_t<InitValueType>, BinaryFunction>

umfranzw committed:
Change device scan default accumulator type

The default accumulator types are changing for ROCm 7.0 as described below.
Note that this is a breaking change.
  * rocprim::inclusive_scan
    * previous default: class AccType = typename std::iterator_traits<InputIterator>::value_type>
    * new default: class AccType = rocprim::invoke_result_binary_op_t<typename std::iterator_traits<InputIterator>::value_type, BinaryFunction>`
  * `rocprim::deterministic_inclusive_scan
    * previous default: class AccType = typename std::iterator_traits<InputIterator>::value_type>`
    * new default: class AccType = rocprim::invoke_result_binary_op_t<typename std::iterator_traits<InputIterator>::value_type, BinaryFunction>`
  * `rocprim::exclusive_scan
    * previous default: class AccType = detail::input_type_t<InitValueType>>
    * new default: class AccType = rocprim::invoke_result_binary_op_t<rocprim::detail::input_type_t<InitValueType>, BinaryFunction>
  * `rocprim::deterministic_exclusive_scan
    * previous default: class AccType = detail::input_type_t<InitValueType>>
    * new default: class AccType = rocprim::invoke_result_binary_op_t<rocprim::detail::input_type_t<InitValueType>, BinaryFunction>

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,28 @@
+
+### Changed
+
+* The default scan accumulator types for device-level scan algorithms has be changed. This is a breaking change.
```
```suggestion
* The default scan accumulator types for device-level scan algorithms have changed. This is a breaking change.
```

umfranzw replied:
Thanks - I've fixed this up in the new patch set.

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,28 @@
+### Changed
+
+* The default scan accumulator types for device-level scan algorithms has be changed. This is a breaking change.
+The previous default accumulator types could lead to situations in which unexpected overflow occured (for example,
```
```suggestion
The previous default accumulator types could lead to situations in which unexpected overflow occured, such as
```

umfranzw replied:
Fixed, thanks.

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,28 @@
+
+* The default scan accumulator types for device-level scan algorithms has be changed. This is a breaking change.
+The previous default accumulator types could lead to situations in which unexpected overflow occured (for example,
+if the input or inital type were smaller than the output type).
```
```suggestion
when the input or inital type was smaller than the output type.
```

umfranzw replied:
Done.

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,28 @@
+The previous default accumulator types could lead to situations in which unexpected overflow occured (for example,
+if the input or inital type were smaller than the output type).
+
+The following is the complete list of affected functions and how their default accumulator types are changing:
```
```suggestion
This is a complete list of affected functions and how their default accumulator types are changing:
```

umfranzw replied:
Done.

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -2,6 +2,28 @@
+The following is the complete list of affected functions and how their default accumulator types are changing:
+  * `rocprim::inclusive_scan`
+    * current default: `class AccType = typename std::iterator_traits<InputIterator>::value_type>`
+    * future default: `class AccType = rocprim::invoke_result_binary_op_t<typename std::iterator_traits<InputIterator>::value_type, BinaryFunction>`
```
Is this change in 7.0 or upcoming?

Because future implies that this is changing in a release *after* 7.0.

If it's changing *after* 7.0, then this should be in the upcoming changes section.

If this is changing *in* 7.0, then this should be:

```suggestion
    * new default: `class AccType = rocprim::invoke_result_binary_op_t<typename std::iterator_traits<InputIterator>::value_type, BinaryFunction>`
```

umfranzw replied:
Nice catch, thanks. Yes, this change is happening in 7.0. I thought I'd fixed this locally, but I must have undone my changes before I pushed. I've changed the descriptors to "past" and "new" to make it clearer.

spolifroni-amd reviewed (changes requested):
The changelogs need some grammatical changes and some clarification.

BTW, if this needs documentation changes, let me know.

NguyenNhuDi reviewed (approved):
lgtm

umfranzw committed:
Grammatical corrections for 7.0 changelog items

This fix corrects some grammatical issues in the changelog.

umfranzw committed:
Update device_scan tests to use new defaults

Update the inclusive and exclusive host scan
routines so they use the new accumulator
type defaults.
